src/Agent.Worker/Handlers/ProcessHandler.cs

```diff
@@ -217,7 +222,11 @@ private void OnOutputDataReceived(object sender, ProcessDataReceivedEventArgs e)
                         if (index > 0)
                         {
                             string key = line.Substring(0, index);
-                            string value = line.Substring(index + 1);
+
+                            if (_readonlyEnvVariables.Contains(key, StringComparer.OrdinalIgnoreCase))
+                            {
+                                return;
```

Comment:
Consider: log an error (or at least a warning) when attempting to set a 'read-only variable'

Reply:
I'm not really sure we want to throw error or warning here, because customer or task owner didn't really control what variables will be. [Here](https://github.com/microsoft/azure-pipelines-agent/pull/4174/files#diff-5a365d72b0ff12989383e1506a6bc0ede99568617da8575934401023b0e7a17cR131) we get all the environment variables and process them from case to case